# Assignment 3 – Tensor Python Tasks


## Task 1 — Create a Tensor
Create a 1D tensor containing 25.2, 25.5, 24.9, 26.1, 25.8.

In [1]:
import numpy as np
sensor = np.array([25.2, 25.5, 24.9, 26.1, 25.8])
print("Tensor:", sensor)
print("Shape:", sensor.shape)
print("Dimensions:", sensor.ndim)
print("Elements:", sensor.size)

Tensor: [25.2 25.5 24.9 26.1 25.8]
Shape: (5,)
Dimensions: 1
Elements: 5


Expected: shape `(5,)`, 1 dimension, 5 elements.

## Task 2 — From Vector to Matrix
Rows = time, columns = sensors.

In [2]:
temperature = [25, 26, 25, 27, 28]
pressure = [101, 102, 101, 103, 104]
humidity = [60, 62, 61, 63, 64]
vibration = [0.2, 0.3, 0.2, 0.4, 0.5]
X = np.array([temperature, pressure, humidity, vibration]).T
print(X)
print("Shape:", X.shape)
print("Dimensions:", X.ndim)
print("First time step:", X[0])
print("Temperature:", X[:, 0])
print("Vibration:", X[:, 3])

[[ 25.  101.   60.    0.2]
 [ 26.  102.   62.    0.3]
 [ 25.  101.   61.    0.2]
 [ 27.  103.   63.    0.4]
 [ 28.  104.   64.    0.5]]
Shape: (5, 4)
Dimensions: 2
First time step: [ 25.  101.   60.    0.2]
Temperature: [25. 26. 25. 27. 28.]
Vibration: [0.2 0.3 0.2 0.4 0.5]


Expected shape `(5, 4)`. Axis 0 is time; axis 1 is sensor.

## Task 3 — Create a 3D Tensor
Interpretation: Machine × Time × Sensor.

In [3]:
X3 = np.arange(10 * 5 * 4).reshape(10, 5, 4)
print("Shape:", X3.shape)

# Machine 3 = index 2, time step 2 = index 1, temperature = sensor index 0
print("Machine 3, time 2, temperature:", X3[2, 1, 0])
print("Machine 5:", X3[4])
print("All machines at time 1:", X3[:, 0, :])
print("Sensor 4 from all machines/times:", X3[:, :, 3])

Shape: (10, 5, 4)
Machine 3, time 2, temperature: 44
Machine 5: [[80 81 82 83]
 [84 85 86 87]
 [88 89 90 91]
 [92 93 94 95]
 [96 97 98 99]]
All machines at time 1: [[  0   1   2   3]
 [ 20  21  22  23]
 [ 40  41  42  43]
 [ 60  61  62  63]
 [ 80  81  82  83]
 [100 101 102 103]
 [120 121 122 123]
 [140 141 142 143]
 [160 161 162 163]
 [180 181 182 183]]
Sensor 4 from all machines/times: [[  3   7  11  15  19]
 [ 23  27  31  35  39]
 [ 43  47  51  55  59]
 [ 63  67  71  75  79]
 [ 83  87  91  95  99]
 [103 107 111 115 119]
 [123 127 131 135 139]
 [143 147 151 155 159]
 [163 167 171 175 179]
 [183 187 191 195 199]]


Shapes: machine 5 `(5,4)`, all machines at time 1 `(10,4)`, sensor 4 `(10,5)`. NumPy uses zero-based indexing.

## Task 4 — Indexing a Tensor

In [4]:
X = np.arange(24).reshape(2, 3, 4)
print(X)
print("X.shape:", X.shape)
print("X[0]:", X[0])
print("X[1]:", X[1])
print("X[0,0]:", X[0,0])
print("X[0,0,0]:", X[0,0,0])

[[[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]]
X.shape: (2, 3, 4)
X[0]: [[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]
X[1]: [[12 13 14 15]
 [16 17 18 19]
 [20 21 22 23]]
X[0,0]: [0 1 2 3]
X[0,0,0]: 0


`X[0]` is the first block `(3,4)`. `X[0,0]` is its first row `(4,)`. `X[0,0,0]` is one scalar.

## Task 5 — Slicing

In [5]:
X = np.arange(60).reshape(3, 4, 5)
a = X[0]
b = X[:, 0, :]
c = X[:, :, 2]
d = X[:, :, -2:]
print("First block:", a.shape)
print("First row from every block:", b.shape)
print("Third element from every row/block:", c.shape)
print("Last two elements:", d.shape)

First block: (4, 5)
First row from every block: (3, 5)
Third element from every row/block: (3, 4)
Last two elements: (3, 4, 2)


Expected shapes: `(4,5)`, `(3,5)`, `(3,4)`, `(3,4,2)`.

## Task 6 — Image Tensor

In [6]:
image = np.random.randint(0, 256, size=(128,128,3), dtype=np.uint8)
red = image[:, :, 0]
green = image[:, :, 1]
blue = image[:, :, 2]
print("Image:", image.shape)
print("Red:", red.shape)
print("Green:", green.shape)
print("Blue:", blue.shape)
print("Max:", image.max())
print("Min:", image.min())

Image: (128, 128, 3)
Red: (128, 128)
Green: (128, 128)
Blue: (128, 128)
Max: 255
Min: 0


Each color channel has shape `(128,128)`. Pixel values range from 0 to 255 in this generated image.

## Task 7 — Batch of Images

In [7]:
images = np.random.rand(32,128,128,3)
first_image = images[0]
first_10 = images[:10]
red_all = images[:, :, :, 0]
pixel_50_50 = images[:, 50, 50, :]
print(images.shape)
print(first_image.shape)
print(first_10.shape)
print(red_all.shape)
print(pixel_50_50.shape)

(32, 128, 128, 3)
(128, 128, 3)
(10, 128, 128, 3)
(32, 128, 128)
(32, 3)


Expected: `(32,128,128,3)`, `(128,128,3)`, `(10,128,128,3)`, `(32,128,128)`, `(32,3)`.

## Task 8 — Reshape

In [8]:
X = np.arange(24)
for shape in [(4,6), (2,3,4), (6,4), (24,)]:
    Y = X.reshape(shape)
    print("\nShape:", Y.shape)
    print("Dimensions:", Y.ndim)
    print(Y)


Shape: (4, 6)
Dimensions: 2
[[ 0  1  2  3  4  5]
 [ 6  7  8  9 10 11]
 [12 13 14 15 16 17]
 [18 19 20 21 22 23]]

Shape: (2, 3, 4)
Dimensions: 3
[[[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]]

Shape: (6, 4)
Dimensions: 2
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]
 [16 17 18 19]
 [20 21 22 23]]

Shape: (24,)
Dimensions: 1
[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23]


Every reshape contains 24 elements. Reshape changes the organization/shape, not the total number of elements.

## Task 9 — Reshape vs Transpose

In [9]:
X = np.arange(24).reshape(2,3,4)
reshaped = X.reshape(3,2,4)
transposed = X.transpose(1,0,2)
print("Reshape:\n", reshaped)
print("Transpose:\n", transposed)
print("Same values in same positions?", np.array_equal(reshaped, transposed))

Reshape:
 [[[ 0  1  2  3]
  [ 4  5  6  7]]

 [[ 8  9 10 11]
  [12 13 14 15]]

 [[16 17 18 19]
  [20 21 22 23]]]
Transpose:
 [[[ 0  1  2  3]
  [12 13 14 15]]

 [[ 4  5  6  7]
  [16 17 18 19]]

 [[ 8  9 10 11]
  [20 21 22 23]]]
Same values in same positions? False


They are not conceptually the same: `reshape` reorganizes elements into a new shape; `transpose` changes axis order.

## Task 10 — Broadcasting

In [10]:
X = np.ones((3,4))
b = np.array([10,20,30,40])
print("X + b:\n", X + b)
print("X * b:\n", X * b)

X + b:
 [[11. 21. 31. 41.]
 [11. 21. 31. 41.]
 [11. 21. 31. 41.]]
X * b:
 [[10. 20. 30. 40.]
 [10. 20. 30. 40.]
 [10. 20. 30. 40.]]


A `(4,)` vector broadcasts across the last dimension of `(3,4)`, so the same four values are applied to every row.

## Task 11 — Broadcasting with Sensor Data

In [11]:
X = np.random.randn(100,10)
calibration = np.array([1.2,0.8,1.1,0.5,0.9,1.0,0.7,1.3,0.6,0.4])
corrected = X - calibration
print("X:", X.shape)
print("Calibration:", calibration.shape)
print("Corrected:", corrected.shape)

X: (100, 10)
Calibration: (10,)
Corrected: (100, 10)


The output remains `(100,10)` because one calibration value is broadcast to each corresponding sensor column.

## Task 12 — Tensor Aggregation

In [12]:
X = np.random.randn(100,20,5)
mean_time = X.mean(axis=1)
mean_sensors = X.mean(axis=2)
mean_machines = X.mean(axis=0)
max_time = X.max(axis=1)
std_time = X.std(axis=1)
print("Mean across time:", mean_time.shape)
print("Mean across sensors:", mean_sensors.shape)
print("Mean across machines:", mean_machines.shape)
print("Max across time:", max_time.shape)
print("Std across time:", std_time.shape)

Mean across time: (100, 5)
Mean across sensors: (100, 20)
Mean across machines: (20, 5)
Max across time: (100, 5)
Std across time: (100, 5)


Expected shapes: mean-time `(100,5)`, mean-sensors `(100,20)`, mean-machines `(20,5)`, max-time `(100,5)`, std-time `(100,5)`.

## Task 13 — Normalize a Tensor

In [13]:
X = np.random.randn(100,20,5)
mean = X.mean(axis=(0,1), keepdims=True)
std = X.std(axis=(0,1), keepdims=True)
X_normalized = (X - mean) / std
print("Normalized shape:", X_normalized.shape)
print("Mean per sensor:", X_normalized.mean(axis=(0,1)))
print("Std per sensor:", X_normalized.std(axis=(0,1)))

Normalized shape: (100, 20, 5)
Mean per sensor: [-1.67608982e-17 -1.95399252e-17 -1.90958360e-17 -5.54001289e-17
 -1.74860126e-17]
Std per sensor: [1. 1. 1. 1. 1.]


After normalization, each sensor should have mean approximately 0 and standard deviation approximately 1.

## Task 14 — Flattening

In [14]:
image = np.random.rand(28,28,3)
flat_image = image.reshape(2352)
batch = np.random.rand(64,28,28,3)
flat_batch = batch.reshape(64,2352)
print("Image:", image.shape, "->", flat_image.shape)
print("Batch:", batch.shape, "->", flat_batch.shape)

Image: (28, 28, 3) -> (2352,)
Batch: (64, 28, 28, 3) -> (64, 2352)


A single image has `28×28×3 = 2352` elements. For a batch, keep the batch axis: `(64,2352)`. Flattening everything to `(150528,)` would lose the separation between the 64 images.

## Task 15 — Concatenation

In [15]:
visual_features = np.random.randn(32,128)
sensor_features = np.random.randn(32,64)
combined = np.concatenate([visual_features, sensor_features], axis=1)
print("Visual:", visual_features.shape)
print("Sensor:", sensor_features.shape)
print("Combined:", combined.shape)

Visual: (32, 128)
Sensor: (32, 64)
Combined: (32, 192)


Result: `(32,192)` because `128+64=192`. Concatenation along the feature axis is possible because both tensors have 32 samples.

## Task 16 — Practical Mini Challenge

In [16]:
X = np.random.randn(500,60,8)

machine_10 = X[9]
first_20 = X[:, :20, :]
sensor_5 = X[:, :, 4]
average_sensor = X.mean(axis=(0,1))
average_time = X.mean(axis=(0,2))

machine_average = X.mean(axis=(1,2))
best_machine = np.argmax(machine_average) + 1

sensor_mean = X.mean(axis=(0,1), keepdims=True)
sensor_std = X.std(axis=(0,1), keepdims=True)
X_normalized = (X - sensor_mean) / sensor_std

X_train = X[:400]
X_test = X[400:]

print("Dataset:", X.shape)
print("Machine 10:", machine_10.shape)
print("First 20 time steps:", first_20.shape)
print("Sensor 5:", sensor_5.shape)
print("Average each sensor:", average_sensor.shape)
print("Average each time step:", average_time.shape)
print("Highest-average machine:", best_machine)
print("Normalized:", X_normalized.shape)
print("Training:", X_train.shape)
print("Testing:", X_test.shape)

Dataset: (500, 60, 8)
Machine 10: (60, 8)
First 20 time steps: (500, 20, 8)
Sensor 5: (500, 60)
Average each sensor: (8,)
Average each time step: (60,)
Highest-average machine: 454
Normalized: (500, 60, 8)
Training: (400, 60, 8)
Testing: (100, 60, 8)


Expected shapes: dataset `(500,60,8)`, machine 10 `(60,8)`, first 20 `(500,20,8)`, sensor 5 `(500,60)`, sensor averages `(8,)`, time averages `(60,)`, training `(400,60,8)`, testing `(100,60,8)`. The highest-average machine depends on random data.

## Task 17 — PyTorch

In [2]:
%pip install torch

  Using cached torch-2.8.0-cp39-cp39-win_amd64.whl (241.2 MB)
  Using cached networkx-3.2.1-py3-none-any.whl (1.6 MB)
  Using cached fsspec-2025.10.0-py3-none-any.whl (200 kB)
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 23.1.2 -> 26.0.1
[notice] To update, run: c:\users\hp\appdata\local\programs\python\python39\python.exe -m pip install --upgrade pip


In [1]:
import torch

X = torch.arange(24).reshape(2,3,4)
print(X)
print("Shape:", X.shape)
print("Dimensions:", X.ndim)
print("Elements:", X.numel())

X = torch.randn(32,100,768)
print("New tensor shape:", X.shape)

tensor([[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],

        [[12, 13, 14, 15],
         [16, 17, 18, 19],
         [20, 21, 22, 23]]])
Shape: torch.Size([2, 3, 4])
Dimensions: 3
Elements: 24
New tensor shape: torch.Size([32, 100, 768])


For `(32,100,768)`, a reasonable interpretation is 32 sequences/samples × 100 tokens/time steps × 768 features/embedding dimensions.